<a href="https://colab.research.google.com/github/WaliMuhammadAhmad/CODEX-Dataset/blob/main/Train4Llama.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import tarfile
import logging
import shutil
import json
import csv
import os

In [2]:
from google.colab import userdata
from huggingface_hub import login
from huggingface_hub import Repository
from huggingface_hub import create_repo

### Define Global Variables

In [18]:
folder = 'train'
files = 0
count = 1

# Path
file = f'/content/drive/MyDrive/Colab Notebooks/Dataset/{folder}.tar.bz2'

input_folder = f'/content/{folder}/'

output_folder = '/content/Train4llama'

### Define Functions

In [4]:
def write_to_csv(output_csv_file, formatted_data, is_first_row=False):
    mode = 'w' if is_first_row else 'a'
    with open(output_csv_file, mode, newline='', encoding='utf-8') as csv_file:
        csv_writer = csv.writer(csv_file)
        if is_first_row:
            csv_writer.writerow(['text'])  # Set the column name as 'text'
        csv_writer.writerow([formatted_data])

In [5]:
def main():
  print(f"Looking into : {input_folder}")
  for foldername in os.listdir(input_folder):
      folder_path = os.path.join(input_folder, foldername)

      if os.path.isdir(folder_path):
          entries = os.listdir(folder_path)
          files = [entry for entry in entries if os.path.isfile(os.path.join(folder_path, entry))]

          for file in files:
              if file.endswith('_corpus.json'):
                  input_file = os.path.join(folder_path, file)
                  input_file_base = input_file.replace('.json', '')
                  output_csv_file = os.path.join(output_folder, f'{os.path.basename(input_file_base)}.csv')
                  # Reset the rows_written counter for each folder
                  rows_written = 0

                  try:
                      with open(input_file, 'r') as f:
                          input_data = json.load(f)
                          input_text = input_data.get("src_fm_fc_ms_ff", "")
                          target_text = input_data.get("target", "")

                      formatted_data = f"<s> <<SYS>> Generate Unit Test Case for this Unit <</SYS>> [INST] {input_text} [/INST] {target_text} </s>"

                      # Write data to the CSV file
                      write_to_csv(output_csv_file, formatted_data, rows_written == 0)
                      rows_written += 1
                  except Exception as e:
                      error_message = f"Error processing file {input_file}: {e}"
                      logger.error(error_message)

In [15]:
def count_files(directory=dir):
  try:
      entries = os.listdir(directory)
      files = [entry for entry in entries if os.path.isfile(os.path.join(directory, entry))]
      print(f"Number of files in '{directory}': {len(files)}")
  except Exception as e:
      logger.error(f"Error: {e}")

  return len(files)

### Start

In [7]:
shutil.rmtree('/content/sample_data')

In [8]:
log_format = "%(asctime)s - %(levelname)s - %(message)s"
logging.basicConfig(filename="logs.log", filemode="w", format=log_format, level=logging.WARNING)

logger = logging.getLogger(__name__)
file_handler = logging.FileHandler('logs.log')
file_handler.setLevel(logging.WARNING)
formatter = logging.Formatter(log_format)
file_handler.setFormatter(formatter)
logger.addHandler(file_handler)

In [9]:
with tarfile.open(file, 'r:bz2') as tar:
  tar.extractall('/content/')

print(f"Extracted {file} @ {input_folder}.")

Extracted /content/drive/MyDrive/Colab Notebooks/Dataset/train.tar.bz2 @ /content/train/.


In [10]:
folders = os.listdir(input_folder)
print(folders)

['11566581', '2139087', '2036666', '2192780', '129501565', '4643211', '75753179', '55209260', '124578706', '77809403', '65041301', '72017294', '23175652', '135418301', '53684862', '246774970', '590532', '93951167', '162913798', '30124869', '1081751', '64405938', '2550825', '23737510', '5402462', '125640293', '19436066', '63452341', '86107810', '146397561', '209204932', '11140459', '2425139', '142141420', '63372616', '53291241', '188429680', '101650961', '10856630', '17621324', '65975333', '145960323', '1031515', '2488113', '166637026', '34336359', '3935296', '7748336', '31564630', '141270248', '7499734', '65973286', '88139536', '11976306', '46204890', '4653065', '3513191', '766240', '155886348', '37007492', '62177795', '59681438', '14285346', '13127985', '48363025', '117353090', '85987914', '170887039', '34387413', '1076632', '15106163', '83457980', '14176513', '71434550', '251068401', '37447498', '1206855', '3249257', '98026470', '39905749', '65266635', '51307016', '30957532', '149970

In [14]:
for f in folders:
  files +=count_files(f'{input_folder}/{f}')

print(f"Total files in {input_folder} : {files}")

Total files in /content/train/ : 624022


In [16]:
userdata.get('huggingface')

'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'

In [17]:
login("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",add_to_git_credential=True)

Token is valid (permission: write).
Your token has been saved in your configured git credential helpers (store).
Your token has been saved to /root/.cache/huggingface/token
Login successful


In [19]:
create_repo("CodexAI/Train4llama", repo_type="dataset", private=True)

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the huggingface Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


RepoUrl('https://huggingface.co/datasets/CodexAI/Train4llama', endpoint='https://huggingface.co', repo_type='dataset', repo_id='CodexAI/Train4llama')

In [20]:
repo_url='https://huggingface.co/datasets/CodexAI/Train4llama'

repo = Repository(local_dir=output_folder, clone_from=repo_url)

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_deprecation.py:131: FutureWarning: 'Repository' (from 'huggingface_hub.repository') is deprecated and will be removed from version '1.0'. Please prefer the http-based alternatives instead. Given its large adoption in legacy code, the complete removal is only planned on next major release.
For more details, please read https://huggingface.co/docs/huggingface_hub/concepts/git_vs_http.
  warnings.warn(warning_message, FutureWarning)
Cloning https://huggingface.co/datasets/CodexAI/Train4llama into local empty directory.


In [21]:
main()

Looking into : /content/train/


In [27]:
print(count_files(output_folder))

Number of files in '/content/Train4llama': 624023
624023


files count in '/content/Train4llama' is one more bcz it is a repo and has a '.gitattribute'  in it which the os function is also counting a files so 624023-1 = 624022 and that's our number!!

### Push to huggingface

push to huggingface repo using legacy git method, this takes tons of time but immune to connection speed

In [28]:
repo.git_pull()

In [29]:
if os.path.exists(output_folder):
  all_files = [f for f in os.listdir(output_folder) if os.path.isfile(os.path.join(output_folder, f))]
  files_per_folder = 10000
  for i in range(0, len(all_files), files_per_folder):
    subfolder_name = f"{count}"
    subfolder_path = os.path.join(output_folder, subfolder_name)
    if not os.path.exists(subfolder_path):
        os.makedirs(subfolder_path)
    for file in all_files[i:i + files_per_folder]:
        shutil.move(os.path.join(output_folder, file), os.path.join(subfolder_path, file))

    count += 1
else:
    print(f"{output_folder} does not exists!")

print("Files have been moved to subfolders successfully.")

Files have been moved to subfolders successfully.


In [30]:
for f in os.listdir(output_folder):
  count_files(f'{output_folder}/{f}')

Number of files in '/content/Train4llama/21': 10000
Number of files in '/content/Train4llama/42': 10000
Number of files in '/content/Train4llama/60': 10000
Number of files in '/content/Train4llama/44': 10000
Number of files in '/content/Train4llama/19': 10000
Number of files in '/content/Train4llama/38': 10000
Number of files in '/content/Train4llama/23': 10000
Number of files in '/content/Train4llama/51': 10000
Number of files in '/content/Train4llama/47': 10000
Number of files in '/content/Train4llama/5': 10000
Number of files in '/content/Train4llama/27': 10000
Number of files in '/content/Train4llama/62': 10000
Number of files in '/content/Train4llama/59': 10000
Number of files in '/content/Train4llama/35': 10000
Number of files in '/content/Train4llama/30': 10000
Number of files in '/content/Train4llama/14': 10000
Number of files in '/content/Train4llama/16': 10000
Number of files in '/content/Train4llama/54': 10000
Number of files in '/content/Train4llama/22': 10000
Number of fil

In [31]:
repo.push_to_hub(commit_message="test uploaded")

Adding files tracked by Git LFS: ['29/163079637_232_corpus.csv', '38/163079637_231_corpus.csv', '39/1057490_985_corpus.csv', '47/163079637_227_corpus.csv', '49/163079637_236_corpus.csv', '50/163079637_240_corpus.csv', '50/16547484_968_corpus.csv', '52/163079637_239_corpus.csv', '52/28738447_364_corpus.csv', '57/163079637_237_corpus.csv', '59/163079637_235_corpus.csv', '59/163079637_238_corpus.csv', '6/17559168_2_corpus.csv', '6/79434101_18_corpus.csv', '60/79060427_1836_corpus.csv', '62/50718765_1_corpus.csv']. This may take a bit of time if the files are large.


Upload file 50/163079637_240_corpus.csv: 100%|##########| 5.33k/5.33k [00:00<?, ?B/s]

Upload file 57/163079637_237_corpus.csv: 100%|##########| 1.72k/1.72k [00:00<?, ?B/s]

Upload file 60/79060427_1836_corpus.csv: 100%|##########| 2.16k/2.16k [00:00<?, ?B/s]

Upload file 50/16547484_968_corpus.csv: 100%|##########| 2.81k/2.81k [00:00<?, ?B/s]

Upload file 52/163079637_239_corpus.csv: 100%|##########| 1.80k/1.80k [00:00<?, ?B/s]

Upload file 62/50718765_1_corpus.csv: 100%|##########| 1.75k/1.75k [00:00<?, ?B/s]

Upload file 47/163079637_227_corpus.csv: 100%|##########| 3.42k/3.42k [00:00<?, ?B/s]

Upload file 59/163079637_238_corpus.csv: 100%|##########| 1.72k/1.72k [00:00<?, ?B/s]

Upload file 49/163079637_236_corpus.csv: 100%|##########| 1.71k/1.71k [00:00<?, ?B/s]

Upload file 39/1057490_985_corpus.csv: 100%|##########| 1.60k/1.60k [00:00<?, ?B/s]

Upload file 29/163079637_232_corpus.csv: 100%|##########| 1.45k/1.45k [00:00<?, ?B/s]

Upload file 59/163079637_235_corpus.csv: 100%|##########| 1.34k/1.34k [00:00<?, ?B/s]

Upload file 52/28738447_364_corpus.csv: 100%|##########| 1.24k/1.24k [00:00<?, ?B/s]

Upload file 38/163079637_231_corpus.csv: 100%|##########| 1.17k/1.17k [00:00<?, ?B/s]

Upload file 6/17559168_2_corpus.csv: 100%|##########| 867/867 [00:00<?, ?B/s]

Upload file 6/79434101_18_corpus.csv: 100%|##########| 597/597 [00:00<?, ?B/s]

To https://huggingface.co/datasets/CodexAI/Train4llama
   1b82304..d01f2ce  main -> main

   1b82304..d01f2ce  main -> main



'https://huggingface.co/datasets/CodexAI/Train4llama/commit/d01f2ce1d2bb66e96897e759c67979782e7b4fb2'

Alternatively, use HFApi to push dataset to huggingface, this is faster then legacy git commands but requires consistant connection speed

In [32]:
from huggingface_hub import HfApi
api = HfApi()

In [33]:
os.chdir(output_folder)

In [ ]:
api.upload_folder(
  folder_path=output_folder,
  repo_id="CodexAI/Train4llama",
  repo_type="dataset",
)

### Release the Space

In [34]:
shutil.rmtree(input_folder)
shutil.rmtree(output_folder)

In [35]:
os.remove('/content/logs.log')

In [36]:
from google.colab import drive
drive.flush_and_unmount()

### All set!